### DISCUSSION TITLE
Trouble with agent  / trouble with generating correct navmesh

### DISCUSSION THREAD

**Discussion by Samsy** [isaac-mason/recast-navigation-js#274] [General]
https://github.com/isaac-mason/recast-navigation-js/assets/3786958/25fc13cf-7063-4e4b-b6c9-eeb87712dcd6

Hey hello there, 
Thanks again for the work done, it's really impressive

I've got a bit of trouble understanding the agents : see video

Sometimes they are not able to find a path, and they just hit continuously the wall, but also sometimes, they fail to go through a tile, like if a tile limit is preventing them to go through.

This is my updateAgentPositions, I'm passing the target position ( the player position ) and the delta

```
   updateAgentPositions(target, dt ){

        this.crowd.timeStep = dt

        this.crowd.update(dt)

        let i = 0

        const targetPosition = this.navMeshQuery.getClosestPoint(target)

        while(i < this.agents.length ){

            const agent = this.agents[i]

            agent.goto(targetPosition);

            const agentPosition = agent.position();

            agent.model3D.position.set(agentPosition.x, agentPosition.y, agentPosition.z)

            i++
        }
    }

```

Also, this is how I'm creating those agents : 

```
createAgent( position = {x: 0, y: 0, z: 0}, radius = 2){

        this.navMeshQuery.getClosestPoint(position);

        /* get a random point around the given position */
        const initialAgentPosition = this.navMeshQuery.getRandomPointAround(position, radius)
    
        const speed = (Math.random() * 10 + 10);
        const agent = this.crowd.addAgent(initialAgentPosition, {
            radius: radius,
            height: radius,
            maxAcceleration: speed * 10,
            maxSpeed: speed,
            collisionQueryRange: 0.5,
            pathOptimizationRange: 0,
            separationWeight: 1.0
          });
        
        agent.model3D = new Mesh(new SphereGeometry(radius,10, 10), new MeshBasicMaterial({ color: Math.random()* 0xffffff, wireframe: true }))

        this.agents.push( agent )

        return agent

    }
```
    
I've been following all the presets, moved things around, but never can find a good path finder solution yet
    
    
    
Also, another question, I'm very often getting those geometry discrepancies on the generated mesh, I don't know if anyone ever bumped into this yet
    
<img width="1231" alt="Screenshot 2024-03-01 at 15 46 28" src="https://github.com/isaac-mason/recast-navigation-js/assets/3786958/6a44d0a9-9d58-4256-9d69-5500eec0e269">

Edit: I did fix the discrepancies a bit by lowering the edge length to the minimum, but I think this is also related to : 

https://github.com/isaac-mason/recast-navigation-js/issues/156

**Comment by isaac-mason** [comment]
Hey @Samsy, if you're comfortable doing so, could you share the model you're using and the code you're generating the navmesh with? Happy to take a look 🙂

Some others have reported similar issues with tiled navmeshes in particular. From the visuals I take it you're using a tiled navmesh?

It's often the case that the issue is actually an issue with the recast/detour c++ libraries themselves. That said - I think it's always good to start here and make sure the javascript bindings aren't the issue before creating noise for the library maintainers.

**Reply by Samsy** [reply]
Yes of course : 

Here is the original glb : 

https://github.com/Samsy/glbs/blob/main/navmesh_source.glb

And there is the parameters : 

```
{
    "borderSize": 0,
    "cs": 0.5,
    "ch": 0.5,
    "walkableSlopeAngle": 80,
    "walkableHeight": 10,
    "walkableClimb": 10,
    "walkableRadius": 1,
    "maxEdgeLen": 1,
    "maxSimplificationError": 1.3,
    "minRegionArea": 8,
    "mergeRegionArea": 20,
    "maxVertsPerPoly": 6,
    "detailSampleDist": 6,
    "detailSampleMaxError": 0,
    "tileSize": 512
}
```

Indeed this is using a tiledmesh !

**Reply by isaac-mason** [reply]
Thanks! I will make some time to take a look soon.

**Reply by Samsy** [reply]
Hey hello @isaac-mason  :) 
Bumping into this again, is there something I can do to help exploring the issue ? 

Thanks a lot

**Reply by isaac-mason** [reply]
Sorry @Samsy, I haven't had time to look into this. I would start by exploring [google discussions](https://groups.google.com/g/recastnavigation) and [github issues/discussions](https://github.com/recastnavigation/recastnavigation) for the upstream recast/detour libraries to see if there is common wisdom for dealing with these issues.

**Reply by Samsy** [reply]
Hey hello @isaac-mason 
Maybe I'm wrong, but I did look at the issues on the github and google discussions you linked, but had no chance at finding any similar ones

Can it be related to the pruning ?

**Reply by isaac-mason** [reply]
Hi Samsy,

I'm not certain whether flood fill pruning would fix this issue, I haven't had time to closely look at your model, apologies. Some related discussion on this if you haven't seen it already: https://github.com/isaac-mason/recast-navigation-js/discussions/151#discussioncomment-7093167

@rob-myers left a good suggestion to avoid calling `agent.goto` every frame. That would be a good first thing to try.

Echoing their comment, if you share a codesandbox / GitHub repo myself or someone else can take a look too 🙂

**Comment by rob-myers** [comment]
Hey @Samsy , it looks like `updateAgentPositions(target, dt )` is being invoked every frame e.g. because you're running `this.crowd.update(dt)`.

If so, you're also invoking [`agent.goto(targetPosition)`](https://github.com/isaac-mason/recast-navigation-js/blob/29d7ed2f6b3757f9ab3d724a3bd7723287e87690/packages/recast-navigation-core/src/crowd.ts#L198) every frame too. Under the hood it is calling [requestMoveTarget](https://recastnav.com/classdtCrowd.html#aa30c34a65535118989ccd86542e96797).

I would only invoke this when the target changes. It might help.

If not, could you provide a repro e.g. by exposing a branch on codesandbox?
